File: Interview-Prep/cpp-oop/01-bank-account/starter.cpp
// BankAccount — fill in the TODOs. Build:
//   g++ -std=c++17 -Wall -Wextra -g starter.cpp -o starter && ./starter
#include <iostream>
#include <stdexcept>
#include <string>

class BankAccount {
public:
    // TODO: constructor(s). Should it be explicit?
    // TODO: deposit(double amount)
    // TODO: withdraw(double amount)
    // TODO: double balance() const
    // TODO: const std::string& owner() const
    // TODO: int transactionCount() const

private:
    // TODO: members (owner name, balance, transaction count)
};

int main() {
    BankAccount acct{"Joseph", 100.0};
    acct.deposit(50.0);
    acct.withdraw(30.0);
    std::cout << acct.owner() << " balance: " << acct.balance()
              << " (" << acct.transactionCount() << " txns)\n";

    try {
        acct.withdraw(1000.0);  // should throw (overdraft)
    } catch (const std::exception& e) {
        std::cout << "Rejected: " << e.what() << "\n";
    }
    return 0;
}

File: Interview-Prep/python-oop/01-bank-account/starter.py
"""BankAccount — fill in the TODOs.  Run: python3 starter.py"""


class InsufficientFunds(Exception):
    pass


class BankAccount:
    def __init__(self, owner, balance=0.0):
        # TODO: store owner, balance, transaction count in "private" attrs
        ...

    # TODO: @property balance  (read-only)
    # TODO: @property owner     (read-only)
    # TODO: @property transaction_count (read-only)

    def deposit(self, amount):
        ...  # TODO: validate positive, update balance + count

    def withdraw(self, amount):
        ...  # TODO: validate positive, reject overdraft, update

    def __repr__(self):
        ...  # TODO


if __name__ == "__main__":
    acct = BankAccount("Joseph", 100.0)
    acct.deposit(50.0)
    acct.withdraw(30.0)
    print(acct)
    try:
        acct.balance = 999  # should raise AttributeError (read-only)
    except AttributeError:
        print("balance is read-only ✓")
    try:
        acct.withdraw(1000.0)
    except InsufficientFunds as e:
        print("Rejected:", e)
